src/cpptools.ts

```diff
@@ -0,0 +1,139 @@
+   */
+  codeModel: cms.CodeModelContent;
+  /**
+   * The contents of the CMakeCache.txt, which also provides supplementary
+   * configuration information.
+   */
+  cache: CMakeCache;
+  /**
+   * The path to `cl.exe`, if necessary. VS generators will need this property
+   * because the compiler path is not available via the `kit` nor `cache`
+   * property.
+   */
+  clCompilerPath?: string|null;
+}
+
+/**
+ * The actual class that provides information to the cpptools extension. See
+ * the `CustomConfigurationProvider` interface for information on how this class
+ * should be used.
+ */
+export class CppConfigurationProvider implements cpt.CustomConfigurationProvider {
+  /** Our name visible to cpptools */
+  readonly name = 'CMake Tools';
+  /** Out extension ID, visible to cpptools */
+  readonly extensionId = 'vector-of-bool.cmake-tools';
+
+  /**
+   * Get the SourceFileConfigurationItem from the index for the given URI
+   * @param uri The configuration to get from the index
+   */
+  private _getConfiguration(uri: vscode.Uri): cpt.SourceFileConfigurationItem|undefined {
+    const norm_path = util.normalizePath(uri.fsPath);
+    return this._fileIndex.get(norm_path);
+  }
+
+  /**
+   * Test if we are able to provide a configuration for the given URI
+   * @param uri The URI to look up
+   */
+  async canProvideConfiguration(uri: vscode.Uri) { return !!this._getConfiguration(uri); }
+
+  /**
+   * Get the configurations for the given URIs. URIs for which we have no
+   * configuration are simply ignored.
+   * @param uris The file URIs to look up
+   */
+  async provideConfigurations(uris: vscode.Uri[]) { return util.dropNulls(uris.map(u => this._getConfiguration(u))); }
+
+  /** No-op */
+  dispose() {}
+
+  /**
+   * Index of files to configurations, using the normalized path to the file
+   * as the key.
+   */
+  private readonly _fileIndex = new Map<string, cpt.SourceFileConfigurationItem>();
+
+  /**
+   * Create a source file configuration for the given file group.
+   * @param grp The file group from the code model to create config data for
+   * @param opts Index update options
+   */
+  private _buildConfigurationData(grp: cms.CodeModelFileGroup, opts: CodeModelParams): cpt.SourceFileConfiguration {
+    // If the file didn't have a language, default to C++
+    const lang = grp.language || 'CXX';
+    // Try the group's language's compiler, then the C++ compiler, then the C compiler.
+    const comp_cache = opts.cache.get(`CMAKE_${lang}_COMPILER`) || opts.cache.get('CMAKE_CXX_COMPILER')
+        || opts.cache.get('CMAKE_C_COMPILER');
+    // Try to get the path to the compiler we want to use
+    const comp_path = comp_cache ? comp_cache.as<string>() : opts.clCompilerPath;
+    if (!comp_path) {
+      rollbar.error('Unable to automatically determine compiler', {lang, fileGroup: grp});
+    }
+    const is_msvc = comp_path && (path.basename(comp_path).toLocaleLowerCase() === 'cl.exe');
```

Comment:
**note:** clang on Windows will prefer msvc mode if a Windows SDK is installed.  In cpptools, we fix this for you automatically if `c_cpp_properties.json` is misconfigured, but I didn't test this for custom configs.  I probably should...